### DISCUSSION TITLE
How to define default frame from within .sil file

### DISCUSSION THREAD

**Discussion by danielrode** [sile-typesetter/sile#1579] [Q&A] [answered]
Hello SILE community! I am new to the software and am trying to figure out how to set the default frame (which is how page margins are achieved) from within the main .sil file. I am not finding this in the manual. I know a separate class file can be used, but for ease of portability, I would like some of my documents to remain as one unit. This is what I have so far:

```
\begin[papersize=letter,class=plain]{document}

\use[module=packages.math]

\begin{script}
local margin = "1in"
class.defaultFrameset = {
  content = {
      left = margin,
      right = "100%pw - left(content)",
      top = margin,
      bottom = "100%ph - top(content)"
    }
}
\end{script}

\set[parameter=document.baselineskip,value=24pt]  % double-space
\set[parameter=document.parskip,value=0pt]  % remove in-between paragraph spacing
\language[main=en]
\font[size=12pt,weight=100,family="Iosevka"]

% Document content:
This is some math:
\math{\sum_{n=1}^\infty \frac{1}{n^2} = \frac{\pi^2}{6}}
... and some more text.
\end{document}
```

This gives me an error:
```
! Unexpected Lua error

error summary:
	Processing at: /usr/local/share/sile/core/sile.lua:363: in <snippet>:
		[[␤class.defaultFrameset = {␤  content = {␤      left = margin,␤      right = "100%pw - left(content)"]] near ./playground.sil:5:1: in \script
	Using code at: [string "..."]:9: attempt to index a nil value (global 'class')
```

**Comment by Omikhleia** [comment, marked as answer]
It seems to me that the **masters** package is what you might be looking for.

It should look like this, using SIL-like syntax when defining the frames.
```
\use[module=packages.masters]
\define-master-template[id=mymaster,first-content-frame=content]{
\frame[id=content,left=1in,right=100%pw - 1in, top=....]
\frame{id=folio, ....]
...
]
\switch-master[id=mymaster]

(Your content)
```

Also, I recommend always defining a "folio" frame, or you'll get into trouble with page numbering, etc.

**Reply by Omikhleia** [reply]
As an aside note, regarding your initial attempt with a Lua script -- There is indeed no global variable "class". The class instance, though, _may_ be accessed from Lua, but I am pretty sure overriding its defaultFrameset property would anyway come too late at this point (the class is likely already initialized, so this would come too late to be taken into account.)

**Reply by danielrode** [reply]
This helped me get a great start. Thank you!

I am running into some trouble with adding a folio frame, but I will focus on the more pressing issue first: setting up the frames this way is preventing new pages from generating.
```
\begin[papersize=letter,class=plain]{document}

\use[module=packages.masters]
\define-master-template[id=mymaster,first-content-frame=content]{
  \frame[id=content,left=1in, right=100%pw-1in, top=1in, bottom=100%pw-1in]
}
\switch-master[id=mymaster]

\language[main=en]
\font[size=12pt,weight=100,family="Iosevka"]

% DOCUMENT CONTENT:
HERE IS SOME TEXT
... AND SOME MORE TEXT.

\end{document}
```

With more content in the content section (beyond what can be fit on one page), the content simply trails off the bottom of the page rather than a new page being generated to contain it. How can I fix this? I want to define the default frame parameters for each new page, not just one frame for the current page.

**Reply by Omikhleia** [reply]
This works for me with SILE v0.14.3, leading to 4 pages (with the default font - I don't have Iosevka installed).

```
\begin[papersize=letter,class=plain]{document}
\use[module=packages.masters]
\use[module=packages.lorem]

\define-master-template[id=mymaster,first-content-frame=content]{
  \frame[id=content,left=1in, right=100%pw-1in, top=1in, bottom=100%ph-1in]
  \frame[id=folio,left=1in,right=100%pw-1in,top=bottom(content)+0.2in,height=0.5in]
}
\switch-master[id=mymaster]
\switch-master-one-page[id=mymaster]

\language[main=en]
\font[size=12pt]

% DOCUMENT CONTENT
\lorem[words=2000]

\end{document}
```

Having to call `\switch-master-one-page` after `\swich-master` is weird... and likely a bug that ought to be reported... Likely these things were little tested (it's far more usual, probably, to derive a custom class and define the default frame set in Lua). EDIT: or perhaps a documentation issue, some may say. It's an issue anyway ;)

**Reply by danielrode** [reply]
This worked great! It also fixed the issues I was having trying to define my own folio frame. Thank you!

**Reply by Omikhleia** [reply]
You are welcome. I opened issue #1580 to report the issue you faced here.
